In [1]:
import pandas as pd
import numpy as np

import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

In [2]:
from keras.optimizers import Adam
from sklearn.model_selection import train_test_split

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/sklearn/utils/validation.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
  LARGE_SPARSE_SUPPORTED = LooseVersion(scipy_version) >= '0.14.0'


In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [6]:
concrete_data_columns = concrete_data.columns

x = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
y = concrete_data['Strength'] # Strength column

In [7]:
x.head(2)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28

In [8]:
y.head(2)

0    79.99
1    61.89
Name: Strength, dtype: float64

In [9]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3)

print('x_train:', x_train.shape, 'y_train:', y_train.shape)
print('x_test:', x_test.shape, 'y_test:', y_test.shape)

x_train: (721, 8) y_train: (721,)
x_test: (309, 8) y_test: (309,)


In [10]:
n_cols = x_test.shape[1]

# Baseline Model

In [11]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error', )
    return model

In [13]:
model = regression_model()

model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=50, verbose=2)

scores = model.evaluate(x_test, y_test, verbose=2)

scores







Train on 721 samples, validate on 309 samples
Epoch 1/50


2023-07-28 11:45:32.011651: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2023-07-28 11:45:32.017652: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2394295000 Hz
2023-07-28 11:45:32.018299: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x560fa0bb9d90 executing computations on platform Host. Devices:
2023-07-28 11:45:32.018343: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
2023-07-28 11:45:32.093455: W tensorflow/compiler/jit/mark_for_compilation_pass.cc:1412] (One-time warning): Not using XLA:CPU for cluster because envvar TF_XLA_FLAGS=--tf_xla_cpu_global_jit was not set.  If you want XLA:CPU, either set that envvar, or use experimental_jit_scope to enable XLA:CPU.  To confirm that XLA is active, pass --vmodule=xla_compilation_cache=1 (as a proper command-line fl

 - 0s - loss: 326005.2669 - val_loss: 250518.8281
Epoch 2/50
 - 0s - loss: 203035.9566 - val_loss: 151109.9858
Epoch 3/50
 - 0s - loss: 120308.3655 - val_loss: 85549.5779
Epoch 4/50
 - 0s - loss: 66036.6277 - val_loss: 43839.7606
Epoch 5/50
 - 0s - loss: 32553.5747 - val_loss: 19778.5702
Epoch 6/50
 - 0s - loss: 14288.4570 - val_loss: 7945.0852
Epoch 7/50
 - 0s - loss: 5841.0666 - val_loss: 3393.1326
Epoch 8/50
 - 0s - loss: 2812.0125 - val_loss: 2060.0059
Epoch 9/50
 - 0s - loss: 1953.6950 - val_loss: 1786.4588
Epoch 10/50
 - 0s - loss: 1747.3354 - val_loss: 1735.5529
Epoch 11/50
 - 0s - loss: 1694.6794 - val_loss: 1711.6823
Epoch 12/50
 - 0s - loss: 1656.4131 - val_loss: 1674.3977
Epoch 13/50
 - 0s - loss: 1617.5017 - val_loss: 1644.4506
Epoch 14/50
 - 0s - loss: 1578.5900 - val_loss: 1608.5100
Epoch 15/50
 - 0s - loss: 1543.9829 - val_loss: 1566.7459
Epoch 16/50
 - 0s - loss: 1505.4133 - val_loss: 1533.9586
Epoch 17/50
 - 0s - loss: 1469.5439 - val_loss: 1502.7341
Epoch 18/50
 - 0s 

716.1646888510695

The mean square error is 716.1646888510695

# Model Using Normalized Data

In [14]:
# Normalize data

x_train_norm = (x_train - x_train.mean()) / x_train.std()
x_test_norm = (x_test - x_test.mean()) / x_test.std()

x_train_norm.head(2)


Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
264 -0.657820            -0.84259  1.059591 -1.039395          0.261093   
353 -0.643859            -0.84259  1.826511 -1.240305          0.912585   

     Coarse Aggregate  Fine Aggregate       Age  
264          1.428370        0.341745 -0.672426  
353          1.010132        0.044251  0.871483

In [15]:
x_test_norm.head(2)

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
729  0.547808           -0.889689 -0.775527  0.543152         -1.029453   
67   0.563210            0.693104 -0.775527  2.338753         -1.029453   

     Coarse Aggregate  Fine Aggregate       Age  
729          0.681237        0.551520  0.677070  
67          -0.566673       -2.321997 -0.618715

In [17]:
n_cols = x_test_norm.shape[1]
n_cols

8

In [18]:

def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error', )
    return model

In [19]:

model = regression_model()


model.fit(x_train_norm, y_train, validation_data=(x_test_norm, y_test), epochs=50, verbose=2)

scores = model.evaluate(x_test_norm, y_test, verbose=2)

scores


Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 1557.1938 - val_loss: 1626.3759
Epoch 2/50
 - 0s - loss: 1543.1761 - val_loss: 1612.5393
Epoch 3/50
 - 0s - loss: 1529.4893 - val_loss: 1598.8850
Epoch 4/50
 - 0s - loss: 1515.8639 - val_loss: 1585.4075
Epoch 5/50
 - 0s - loss: 1502.3188 - val_loss: 1571.6196
Epoch 6/50
 - 0s - loss: 1488.5348 - val_loss: 1557.6262
Epoch 7/50
 - 0s - loss: 1474.6080 - val_loss: 1543.1420
Epoch 8/50
 - 0s - loss: 1460.0515 - val_loss: 1528.4787
Epoch 9/50
 - 0s - loss: 1445.2643 - val_loss: 1512.9425
Epoch 10/50
 - 0s - loss: 1429.6774 - val_loss: 1496.7848
Epoch 11/50
 - 0s - loss: 1413.6896 - val_loss: 1479.3895
Epoch 12/50
 - 0s - loss: 1396.7530 - val_loss: 1461.8267
Epoch 13/50
 - 0s - loss: 1379.2767 - val_loss: 1443.1572
Epoch 14/50
 - 0s - loss: 1360.9053 - val_loss: 1423.8308
Epoch 15/50
 - 0s - loss: 1341.7027 - val_loss: 1402.9110
Epoch 16/50
 - 0s - loss: 1321.2417 - val_loss: 1381.8041
Epoch 17/50
 - 0s - loss: 1300.3398

506.5680151880752

The error using the nomalized data is 506.5680151880752 and it is lower when compared to the error obatained when is data is not nomalized (716.1646888510695)

# Increasing the Number of Epoch

In [20]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error', )
    return model

In [21]:
model = regression_model()


model.fit(x_train_norm, y_train, validation_data=(x_test_norm, y_test), epochs=100, verbose=2)

scores = model.evaluate(x_test_norm, y_test, verbose=2)

scores

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 0s - loss: 1531.2659 - val_loss: 1604.3729
Epoch 2/100
 - 0s - loss: 1517.0250 - val_loss: 1590.2188
Epoch 3/100
 - 0s - loss: 1503.2528 - val_loss: 1576.7907
Epoch 4/100
 - 0s - loss: 1490.0761 - val_loss: 1563.3458
Epoch 5/100
 - 0s - loss: 1476.7791 - val_loss: 1550.0481
Epoch 6/100
 - 0s - loss: 1463.3222 - val_loss: 1536.4178
Epoch 7/100
 - 0s - loss: 1449.2375 - val_loss: 1522.1115
Epoch 8/100
 - 0s - loss: 1434.3259 - val_loss: 1507.0422
Epoch 9/100
 - 0s - loss: 1418.6914 - val_loss: 1490.7752
Epoch 10/100
 - 0s - loss: 1401.9038 - val_loss: 1473.9226
Epoch 11/100
 - 0s - loss: 1384.6447 - val_loss: 1455.7891
Epoch 12/100
 - 0s - loss: 1366.3413 - val_loss: 1436.9480
Epoch 13/100
 - 0s - loss: 1346.9588 - val_loss: 1417.5707
Epoch 14/100
 - 0s - loss: 1327.2000 - val_loss: 1396.3564
Epoch 15/100
 - 0s - loss: 1305.8559 - val_loss: 1374.6690
Epoch 16/100
 - 0s - loss: 1283.8211 - val_loss: 1351.5250
Epoch 17/100
 - 0s 

155.73840183887665

The error obtained when the numbr of epochs is increased to 100 is 155.73840183887665 and it is lower than that obtained when 50 epoch was used (506.5680151880752)

# Increased the Number of Hidden Layers

In [22]:
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error', )
    return model

In [24]:
model = regression_model()


model.fit(x_train_norm, y_train, validation_data=(x_test_norm, y_test), epochs=50, verbose=2)

scores = model.evaluate(x_test_norm, y_test, verbose=2)

scores

Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 1530.9762 - val_loss: 1599.4394
Epoch 2/50
 - 0s - loss: 1511.4102 - val_loss: 1576.2255
Epoch 3/50
 - 0s - loss: 1481.2537 - val_loss: 1536.1280
Epoch 4/50
 - 0s - loss: 1427.6911 - val_loss: 1465.8441
Epoch 5/50
 - 0s - loss: 1338.5868 - val_loss: 1349.3465
Epoch 6/50
 - 0s - loss: 1194.9262 - val_loss: 1167.6690
Epoch 7/50
 - 0s - loss: 986.1784 - val_loss: 910.0517
Epoch 8/50
 - 0s - loss: 711.2222 - val_loss: 610.4427
Epoch 9/50
 - 0s - loss: 451.4625 - val_loss: 361.3163
Epoch 10/50
 - 0s - loss: 283.9640 - val_loss: 249.3612
Epoch 11/50
 - 0s - loss: 226.9914 - val_loss: 215.4335
Epoch 12/50
 - 0s - loss: 209.1085 - val_loss: 202.4006
Epoch 13/50
 - 0s - loss: 200.1246 - val_loss: 193.4802
Epoch 14/50
 - 0s - loss: 193.0625 - val_loss: 187.4248
Epoch 15/50
 - 0s - loss: 187.9916 - val_loss: 182.0450
Epoch 16/50
 - 0s - loss: 182.6191 - val_loss: 178.4593
Epoch 17/50
 - 0s - loss: 177.9538 - val_loss: 174.3583

125.23856198209003

Compared to the error obtained when 50 epochs was used (506.5680151880752); when the number of hidden layers is increased to 3, the error obtaied is lesser with a value of 125.23856198209003 